### ISSUE TITLE
Issue with relationship matchers

### ISSUE THREAD

**Issue by mkrakowitzer** [rodjek/rspec-puppet#323]
This relates to: https://tickets.puppetlabs.com/browse/MODULES-2374
but I can reproduce this issue with many builtin types too: 

I get the below error when testing relationships on certain types:

```
     Failure/Error: it do should contain_file('/opt/deploy').that_requires('User[deploy]')
     NoMethodError:
       undefined method `[]' for nil:NilClass
```

```puppet
    user { 'deploy':
      ensure => present,
    }

    file { '/opt/deploy':
      ensure  => 'directory',
      group   => 'deploy',
      require => User['deploy'],
    }
```

when I remove the group parameter, it works fine:
```
Finished in 0.54686 seconds (files took 1.72 seconds to load)
6 examples, 0 failures
```

```puppet
    user { 'deploy':
      ensure => present,
    }

    file { '/opt/deploy':
      ensure  => 'directory',
      require => User['deploy'],
    }
```

or when I add the below it also works fine.

```pupppet
group { 'deploy': }
```

But I should not be required to explicitly define the group.

**Comment by DavidS** [comment]
@raphink recently implemented recursive dependency evaluation, but that is not yet released. Would you mind testing this against the master branch?

**Comment by mkrakowitzer** [comment]
This is against the master branch.

**Comment by raphink** [comment]
@DavidS We've actually experienced the same bug yesterday using the master branch.

See https://github.com/raphink/puppet-freeradius/blob/master/spec/defines/freeradius__mod_spec.rb#L32 and https://github.com/raphink/puppet-freeradius/blob/master/spec/defines/freeradius__mod_spec.rb#L60

Run the `bundle exec rake spec` for this puppet-freeradius module and you'll see the same bug.

**Comment by raphink** [comment]
I don't exclude that it actually originates from my code of course :-)

**Comment by raphink** [comment]
@mkrakowitzer's comment on removing the `group` attribute to make it work gives a hint that it might come from the autorequire bit of the code, since there is an autorequirement on group resources in files.

**Comment by raphink** [comment]
So I'm thinking it could come from https://github.com/rodjek/rspec-puppet/commit/9c295b5de726128bc08acfb1efaeaf5987174563

**Comment by raphink** [comment]
I've managed to reproduce that bug with https://github.com/raphink/rspec-puppet/commit/fc42445a89a367dfc2a055138e062f25b24512c4

**Comment by raphink** [comment]
It seems to be linked to autorequires pointing to resources that don't exist in the catalog. So when checking for autorequires, it should gently continue when the resource is not in the catalog.

**Comment by raphink** [comment]
@DavidS got a fix in #324. Do you want to test that @mkrakowitzer ?

**Comment by raphink** [comment]
https://github.com/raphink/puppet-freeradius/commit/64813590f73133a0d3ebb888eee91ebf802d02a5 did fix it for me (removed two pending)

**Comment by mkrakowitzer** [comment]
+1 works for this issue, as well as the issue I raised here: https://tickets.puppetlabs.com/browse/MODULES-2374

**Comment by raphink** [comment]
Great! Thanks for the feedback @mkrakowitzer

**Comment by DavidS** [comment]
Fixed by #324 ; Thanks everyone involved!
